internal/provider/cloudinit_config.go

```diff
@@ -0,0 +1,173 @@
+package provider
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/textproto"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-provider-cloudinit/internal/hashcode"
+)
+
+// Model and functionality of data source and resource are equivalent.
+type configModel struct {
+	ID           types.String      `tfsdk:"id"`
+	Parts        []configPartModel `tfsdk:"part"`
+	Gzip         types.Bool        `tfsdk:"gzip"`
+	Base64Encode types.Bool        `tfsdk:"base64_encode"`
+	Boundary     types.String      `tfsdk:"boundary"`
+	Rendered     types.String      `tfsdk:"rendered"`
+}
+
+type configPartModel struct {
+	ContentType types.String `tfsdk:"content_type"`
+	Content     types.String `tfsdk:"content"`
+	FileName    types.String `tfsdk:"filename"`
+	MergeType   types.String `tfsdk:"merge_type"`
+}
+
+func validateConfigModel(config configModel) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if config.Gzip.IsUnknown() || config.Base64Encode.IsUnknown() {
+		return diags
+	}
+	setDefaultValues(&config)
+
+	if config.Gzip.ValueBool() && !config.Base64Encode.ValueBool() {
+		diags.AddAttributeError(
+			path.Root("base64_encode"),
+			"Invalid Attribute Configuration",
+			"Expected base64_encode to be set to true when gzip is true.",
+		)
+	}
+
+	return diags
+}
+
+func updateConfigModel(ctx context.Context, config *configModel) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	// v2.2.0 doesn't actually set default values in state properly, so we need to make sure
+	// that we don't use any known empty values from previous versions of state
+	setDefaultValues(config)
```

Comment:
The old provider `v2.2.0` doesn't actually set a default value for `content_type` and just [updates it later](https://github.com/hashicorp/terraform-provider-cloudinit/blob/0cee1aa6ede7b9c9db16d515e6e82aa91c11242d/internal/provider/datasource_cloudinit_config.go#L169) to produce the `rendered` output.

I noticed this inconsistency when I was running the migration tests, here is an example:

## Config (just part block)
![Pasted image 20230210162722](https://user-images.githubusercontent.com/8650838/218206028-87eb6830-e3f6-4344-bc1a-252e6a8c5d90.png)

## v2.2.0 state
![Pasted image 20230210162805](https://user-images.githubusercontent.com/8650838/218206115-6a996ab0-c0db-4184-862a-0d95cd485499.png)

I'm not sure the best way to handle this, but I added this line here to ensure that state gets the proper defaults set and that `rendered` output is not modified.

Reply:
Fixing the state to include the default value seems the best course of action here, on the off-chance a practitioner happens to try and use the `content_type` in a configuration reference.

Reply:
cool, I'll leave it as is then 👍🏻